File: chapter12/fmat.ml
type float_matrix
external fmat_of_array : float array array -> float_matrix = "array_to_fmat"
external array_of_fmat : float_matrix -> float array array = "fmat_to_array"
external fmat_add      : float_matrix -> float_matrix -> float_matrix = "fmat_add"
external fmat_mul      : float_matrix -> float_matrix -> float_matrix = "fmat_mul";;

let assert_true res testname =
  if res then
    print_string ("Test " ^ testname ^ " passed\n")
  else
    print_string ("Test " ^ testname ^ " failed\n");;

let test1 () =
    let fmat = fmat_of_array [| [|2.;3.;-1.;0.|]; [|-7.;2.;1.;10.|] |] in
    let faa = array_of_fmat fmat in
        assert_true (faa = [| [|2.;3.;-1.;0.|]; [|-7.;2.;1.;10.|] |]) "fmat_of_array/array_of_fmat";;

let test2() =
    let m1 = fmat_of_array [|[|1.2;2.3;3.4|]; [|4.5;5.6;6.7|]|] in
    let m2 = fmat_of_array [|[|1.2;2.3;3.4|]; [|4.5;5.6;6.7|]|] in
    let r = fmat_add m1 m2 in
    let faa = array_of_fmat r in
        assert_true (faa = [|[|2.4; 4.6; 6.8|]; [|9.; 11.2; 13.4|]|]) "fmat_add";;

let test3() = 
    let m1 = fmat_of_array [| [|2.;3.;-1.;0.|]; [|-7.;2.;1.;10.|] |] in
    let m2 = fmat_of_array [| [|3.;4.|]; [|2.;1.|]; [|-1.;2.|]; [|2.;7.|] |] in
    let r = fmat_mul m1 m2 in
    let faa = array_of_fmat r in
        assert_true (faa = [|[|13.; 9.|]; [|2.; 46.|]|]) "fmat_mul";;

let () =
    test1();
    test2();
    test3();;

File: chapter12/fmat.c
#include "fmat.h"

value array_to_fmat (value faa)
{
    int   i,j;
    value row;

    value fmat = make_fmat(faa);

    float_matrix_t * fm = (float_matrix_t *) Field(fmat,1);

    for (i=0; i<fm->rows; i++)
    {
        row = Field(faa,i);
        for (j=0; j<fm->cols; j++)
        {
            set(fm, i, j, Double_field(row, j));
        }
    } 
    return fmat;
}

value fmat_to_array (value fmat)
{
    value block, row;
    int i, j;

    float_matrix_t * fm = (float_matrix_t *) Field(fmat,1);

    block = make_faa (fm->rows, fm->cols);

    for (i=0; i<fm->rows; i++)
    {
        row = Field(block, i);
        for (j=0; j<fm->cols; j++)
        {
            Store_double_field(row, j, get(fm, i, j));
        }

    }
        
    return block;
}

value fmat_add (value mat1, value mat2)
{
    int i,j;
    value block;
    double add_val;
    float_matrix_t *m1, *m2, *result;

    m1 = (float_matrix_t *)Field(mat1, 1);
    m2 = (float_matrix_t *)Field(mat2, 1);

    if ((m1->rows != m2->rows) && (m1->cols != m2->cols))
        failwith ("fmat_add: Incompatible matrices");

    block = make_fmat2 (m1->rows, m1->cols);
    result = (float_matrix_t *)Field(block, 1);

    for (i=0; i<m1->rows; i++)
    {
        for (j=0; j<m1->cols; j++)
        {
            add_val = get(m1, i, j) + get(m2, i, j);
            set (result, i, j, add_val);
        }
    }

    return block;
}

value fmat_mul (value mat1, value mat2)
{
    int i,j,k;
    value block;
    double part;
    float_matrix_t *m1, *m2, *result;

    m1 = (float_matrix_t *)Field(mat1, 1);
    m2 = (float_matrix_t *)Field(mat2, 1);

    if (m1->cols != m2->rows)
        failwith ("fmat_mul: Incompatible matrices");

    block = make_fmat2 (m1->rows, m2->cols);
    result = (float_matrix_t *)Field(block, 1);

    for (i=0; i<result->rows; i++)
    {
        for (j=0; j<result->cols; j++)
        {
            part = 0.0;
            for (k=0; k<m1->cols; k++)
            {
                part = part + (get(m1, i, k) * get(m2, k, j));
            }
            set(result, i, j, part);
        }
    }

    return block;
}

double get(float_matrix_t* fmat, int r, int c)
{
    double ret;
    int index;

    if ((r > fmat->rows) || (c > fmat->cols))
        failwith ("get: Index out-of-bound");

    index = r*fmat->cols + c;
    ret = fmat->data[index];
    return ret;
}

void set(float_matrix_t* fmat, int r, int c, double val)
{
    int index;

    if ((r > fmat->rows) || (c > fmat->cols))
        failwith ("set: Index out-of-bound");

    index = r*fmat->cols + c;
    fmat->data[index] = val;
    return;
}

value make_fmat (value faa)
{
    value block;
    int rows, cols;

    rows=Wosize_val(faa);
    cols=(rows > 0) ? (Wosize_val(Field(faa,0))) : 0;

    block = make_fmat2 (rows, cols);
    return block;
}

value make_fmat2(int rows, int cols)
{
    value block;
    int   use;
    float_matrix_t *fmat;

    fmat = malloc(sizeof(float_matrix_t));

    if (fmat == NULL)
        failwith ("make_fmat2: Failed to allocate float_matrix");

    fmat->rows = rows;
    fmat->cols = cols;

    fmat->data = (double *)malloc(sizeof(double) * rows * cols);
    if (fmat->data == NULL)
        failwith ("make_fmat2: Failed to allocate float_matrix data");

    use = (sizeof(double)*rows*cols) + sizeof(float_matrix_t);
    block = alloc_final(2, fin_fmat, use, 1000000);
    Field(block,1)=(value)fmat;
    return block;
}

value make_faa(int rows, int cols)
{
    value block;
    int i, num_elems;

    num_elems = cols * Double_wosize;

    block = alloc_tuple(rows);

    for (i=0; i<rows; i++)
    {
        Field(block, i) = alloc (num_elems, Double_array_tag);
    }

    return block;
}

void  fin_fmat (value v)
{
    float_matrix_t *fmat = (float_matrix_t *) Field(v,1);

    free (fmat->data);
    free (fmat);
    //printf("float_matrix gc'd\n");
    //fflush(stdout);
}












